pkg/operator/controller/canary/controller.go

```diff
@@ -0,0 +1,174 @@
+		return nil, err
+	}
+
+	return c, nil
+}
+
+// Reconcile ensures that the canary controller's resources
+// are in the desired state.
+func (r *reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	result := reconcile.Result{}
+	errors := []error{}
+
+	if err := r.ensureCanaryNamespace(); err != nil {
+		// Return if the canary namespace cannot be created since
+		// resource creation in a namespace that does not exist will fail.
+		return result, fmt.Errorf("failed to ensure canary namespace: %v", err)
+	}
+
+	if err := r.ensureCanaryRole(); err != nil {
+		errors = append(errors, fmt.Errorf("failed to ensure canary role: %v", err))
+	}
+
+	haveDs, daemonset, err := r.ensureCanaryDaemonSet()
+	if err != nil {
+		errors = append(errors, fmt.Errorf("failed to ensure canary daemonset: %v", err))
+	}
+	if !haveDs {
+		errors = append(errors, fmt.Errorf("failed to get canary daemonset: %v", err))
+	}
+
+	trueVar := true
+	daemonsetRef := metav1.OwnerReference{
+		APIVersion: "apps/v1",
+		Kind:       "daemonset",
+		Name:       daemonset.Name,
+		UID:        daemonset.UID,
+		Controller: &trueVar,
+	}
+
+	haveService, service, err := r.ensureCanaryService(daemonsetRef)
+	if err != nil {
+		errors = append(errors, fmt.Errorf("failed to ensure canary service: %v", err))
+	}
+	if !haveService {
+		errors = append(errors, fmt.Errorf("failed to get canary service: %v", err))
+	}
+
+	if haveRoute, _, err := r.ensureCanaryRoute(service); err != nil {
+		errors = append(errors, fmt.Errorf("failed to ensure canary route: %v", err))
+	} else if !haveRoute {
+		errors = append(errors, fmt.Errorf("failed to get canary route: %v", err))
+	}
+
+	return result, utilerrors.NewAggregate(errors)
+}
+
+// Config holds all the things necessary for the controller to run.
+type Config struct {
+	Namespace   string
+	CanaryImage string
+}
+
+// reconciler handles the actual canary reconciliation logic in response to
+// events.
+type reconciler struct {
+	Config
+
+	client client.Client
+}
+
+// TODO: Canary Controller Phase 2
+// Add callers for these 2 functions
+//
+// Switch the current RoutePort that the route points to.
+// Use this function to periodically update the canary route endpoint
+// to verify if the router has wedged.
+func (r *reconciler) rotateRouteEndpoint(service *corev1.Service, current *routev1.Route) (*routev1.Route, error) {
+	updated, err := cycleServicePort(service, current)
+	if err != nil {
+		return nil, fmt.Errorf("failed to rotate route port: %v", err)
+	}
+
+	_, err = r.updateCanaryRoute(current, updated)
+	if err != nil {
+		return current, err
+	}
+
+	return updated, nil
+}
+
+// cycleServicePort returns a route resource with Spec.Port set to the
+// next available port in service.Spec.Ports that is not the current route.Spec.Port.
+func cycleServicePort(service *corev1.Service, route *routev1.Route) (*routev1.Route, error) {
+	servicePorts := service.Spec.Ports
+	currentPort := route.Spec.Port
```

Comment:
Add a null check so the operator doesn't dereference a null pointer if someone deletes the port on the route.

Reply:
Fixed, thanks!!